### DISCUSSION TITLE
I2S Output of 4 Channels with an ESP32

### DISCUSSION THREAD

**Discussion by bbaldach** [pschatzmann/arduino-audio-tools#1536] [Q&A] [answered]
Hi , I’ve try to implement the "I2S Output of 4 Channels with an ESP32" with 2 DAC (PCM5102A) and an ESP32 DEVKIT V1.
The first DAC is connected to
config1.pin_bck = 25;
config1.pin_ws = 33;
config1.pin_data = 32
and the second one to
config2.pin_bck = 14;
config2.pin_ws = 15;
config2.pin_data = 22;

but I only have sound (with a lot of glitch on DAC1)

In the terminal I have that “error” : [W] AudioOutput.h : 843 – Could not write all samples

Do you have an idea ?

thank you

```
#include <SPI.h>
#include <SD.h>
#include "AudioTools.h"
#include "AudioCodecs/CodecMP3Helix.h"

const int chipSelect=10;
AudioInfo info(44100, 2, 16);
I2SStream i2s_1; // final output
I2SStream i2s_2; // final out
ChannelsSelectOutput out;
WAVDecoder wav;
EncodedAudioOutput decoder(&out, &wav); // Decoding stream
StreamCopy copier; 
File audioFile;

void setup(){
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);  

  // setup file
  SD.begin(chipSelect);
  audioFile = SD.open("/4.wav");

  // setup i2s
  auto config1 = i2s_1.defaultConfig(TX_MODE);
  config1.copyFrom(info);
  config1.port_no = 0;
  config1.pin_bck = 25;// config1.pin_bck = 14;
  config1.pin_ws = 33;// config1.pin_ws = 15;
  config1.pin_data = 32;// config1.pin_data = 22;  
  config1.set(info);
  i2s_1.begin(config1);
  
  auto config2 = i2s_2.defaultConfig(TX_MODE);
  config2.copyFrom(info);
  config2.port_no = 1;
  config2.pin_bck = 14;//config2.pin_bck = 25;
  config2.pin_ws = 15;//config2.pin_ws = 33;
  config2.pin_data = 22;//config2.pin_data = 32;
  i2s_2.begin(config2);

  // split channels to different i2s ports
  out.addOutput(i2s_1, 0, 1);
  out.addOutput(i2s_2, 2, 3);
  out.addNotifyAudioChange(i2s_1);
  out.addNotifyAudioChange(i2s_2);
  out.begin(info);

  // setup decoder
  decoder.begin();

  // begin copy
  copier.begin(decoder, audioFile);
}

void loop(){
  if (!copier.copy()) {
    stop();
  }
}
```

**Comment by pschatzmann** [comment, marked as answer]
There have been a couple of improvements & corrections.
Try again with the latest version...

ps. does your input data really have 4 channels ? if it is a regular stereo file you could do 
  out.addOutput(i2s_1, 0, 1);
  out.addOutput(i2s_2, 0, 1);
or
  out.addOutput(i2s_1, 0, 0); // only left on 1
  out.addOutput(i2s_2, 1, 1); // only right on 2

There should be no need for calling addNotifyAudioChange any more

**Reply by bbaldach** [reply]
Ok . I think I have the latest version but I'll double check.

It's a 4 channels Wave file
![image](https://github.com/pschatzmann/arduino-audio-tools/assets/50704114/e44754a0-d196-4f3d-9a5c-048785579284)

**Comment by bbaldach** [comment]
😔 DIN and LCK was inverted on my board ...... 
It works now. 
Only glitch on both DAC but I think it's due to SD card speed.
Thanks a lot !!

**Reply by bbaldach** [reply]
I have not yet found the solution to Glitch ... and my wave file looks played half speed.
How the libray manage 16 bits I2S stream to 32 bits I2S DAC ? (PCM5102A)

**Reply by pschatzmann** [reply]
There is a chapter in the Wiki for format conversion!

**Comment by bbaldach** [comment]
Hi @pschatzmann ,

I've made several test with SPI SD, SDMMC 1 bit and 4 bit, chnage of SD Card, etc...  and I don't found why my file continu to be played "slowly" and with some glitch.

this is my code (very close to your example. Any idea ? 
And this is the wav file used (rename as txt because wav are not allowed)
[4.txt](https://github.com/user-attachments/files/15520745/4.txt)


`#include <FS.h>
#include <SD_MMC.h>
#include <AudioTools.h>
#include <AudioCodecs/CodecWAV.h>

int clk = 36;
int cmd = 35;
int d0 = 37;
int d1 = 20;
int d2 = 41;
int d3 = 42;

AudioInfo info(44100, 2, 16);
I2SStream i2s_1; // final output port 0
I2SStream i2s_2; // final output port 1
ChannelsSelectOutput out;
WAVDecoder wav;
EncodedAudioOutput decoder(&out, &wav); // Decoding stream
StreamCopy copier; 
File audioFile;

void setup(){
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);  

  // setup i2s
  auto config1 = i2s_1.defaultConfig(TX_MODE);
  config1.copyFrom(info);
  config1.port_no = 0;
  config1.pin_bck = 40;
  config1.pin_ws = 38;
  config1.pin_data = 39;
  i2s_1.begin(config1);
  auto config2 = i2s_2.defaultConfig(TX_MODE);
  config2.copyFrom(info);
  config2.port_no = 1;
  config2.pin_bck = 45;
  config2.pin_ws = 21;
  config2.pin_data = 47;
  i2s_2.begin(config2);

    if(! SD_MMC.setPins(clk, cmd, d0,d1,d2,d3)){
         Serial.println("Pin change failed!");
        return;
    }

    // open sdmmc in 4 bit mode
  if (!SD_MMC.begin("/sdcard",false,false,40000,5)){
    Serial.println("SD_MMC Error");
    stop();
  }
  // open file
  audioFile = SD_MMC.open("/4.wav");
  if (!audioFile){
    Serial.println("File does not exist");
    stop();
  }

  // split channels to different i2s ports
  out.addOutput(i2s_1, 0, 1);
  out.addOutput(i2s_2, 2, 3);
  //out.addNotifyAudioChange(i2s_1);
  //out.addNotifyAudioChange(i2s_2);
  out.begin(info);

  // setup decoder
  decoder.begin();

  // begin copy
  copier.begin(decoder, audioFile);
}

void loop(){
  if (!copier.copy()) {
    stop();
  }
}`

**Comment by pschatzmann** [comment]
Did you already try out [this optimization](https://github.com/pschatzmann/arduino-audio-tools/commit/5e1cd86512c6cd0fb3bdb03fb620256d651ed7bb) ?

**Reply by bbaldach** [reply]
Yes, I've try but no result till I comment :
`//AudioLogger::instance().begin(Serial, AudioLogger::Info);  `
Now it works properly

**Reply by bbaldach** [reply]
Looks like AudioLogger consume too much  ressources of the ESP32-S3

**Reply by pschatzmann** [reply]
What makes you thinks so ? 
Did you use the recommended Warning level ?

**Reply by bbaldach** [reply]
> What makes you thinks so ? Did you use the recommended Warning level ?

Because when I comments the audiologger line sound reproduction is correct.
I have used the example who define an "Info" Level. I have to test with "Warning" level to seen if it works correctly.

**Comment by pschatzmann** [comment]
Commenting it out is the same as setting it to Warning...

**Comment by bbaldach** [comment]
I've almost finished my program that play stereo wave, 4 channels wave and MP3 (by switching of decoder on the fly). 
To avoid persistent sounds at the end of the file (I can't be sure that my file ends with zeros), I've added writeSilence in the copier loop.


```
void loop() {
  if (Etat == 1) {

    if (!copier.copy()) {  
      i2s_1.writeSilence(i2s_1.availableForWrite());
      i2s_2.writeSilence(i2s_2.availableForWrite());
    }
  }
}
```
It works good for MP3 and stereo wav but not for 4 channels wave.

Is it a good way to do it ?
Why it doesn't work with 4 channels wave ?
Is existing another way to do it ?
Thanks

**Reply by pschatzmann** [reply]
I don't see any reason why this should not work however I also do not understand why this should be necessary.

Some other ideas to address the issue:
- call end()
- some DACs can be controlled with an enable pin, so set this to low
- make sure that auto_clear is set to true
- make sure that you use an actual i2s core version

**Reply by bbaldach** [reply]
Thanks for your feedback.

- call end() **--> for which "object"? I2S ?**
- some DACs can be controlled with an enable pin, so set this to low **--> It's an idea**
- make sure that auto_clear is set to true **--> It solve the problem for mp3 and stereo wave but always the same for 4 channels wave**
- make sure that you use an actual i2s core version **--> what do you mean by "i2s core version" ?**

**Comment by pschatzmann** [comment]
yes: end on I2SStream
My mistake: I meant ESP32 core version

**Reply by bbaldach** [reply]
So I confirm I use an ESP32 core version.
By only using `i2s.end()` it doesn't have effect but by adding `out.writeSilence(out.availableForWrite());` It looks to works fine.

`if (!copier.copy()) {`
`i2s_1.end();`
`i2s_2.end();`
`out.writeSilence(out.availableForWrite());`
`}`

**Comment by pschatzmann** [comment]
Just for your information so that you can double check what you are using: the currently latest ESP32 core version is 3.0.1!

Strange: After calling end() any subsequent output will not do anything at all...

**Reply by bbaldach** [reply]
In confirm, I use the 3.0.1
![image](https://github.com/pschatzmann/arduino-audio-tools/assets/50704114/7057aba4-b920-454d-9dce-8feb17ad660e)

Yes I know, it was also my understanding . Maybe a side effect of the  ChannelsSelectOutput ?

**Reply by daviti231** [reply]
I am trying to connect one ADC I2S ADC and two DAC HiLetgo PCM5102,
As I understand there is only 2 port available on ESP32, so my code is that 1st port I am using with shared BCK and WS pins for both ADC and DAC, 2st port I am using only for DAC, problem is that I only get sound in 2 st dac, 1st dac is giving me very noise and unclear sound.

the code looks like this, what i am doing wrong :
#include "AudioTools.h"
#include "AudioLibs/AudioFaust.h"
#include "volumes.h"


const uint16_t sample_rate = 44100;
const uint16_t bits_per_sample = 44100;
// Define the audio configurations
AudioInfo info(sample_rate , 2, 16);  // For 2-channel output
AudioInfo info4(sample_rate , 4, 16); // For 4-channel output

// Define the audio components

mydsp dsp;  // Faust DSP object

I2SStream i2s_1;  // First I2S output stream
I2SStream i2s_2;  // Second I2S output stream
I2SStream adc;    // I2S input stream (ADC)

ChannelsSelectOutput out;  // Output channel selector
// Faust stream using the output channel selector
StreamCopy copier(out, adc);  // Copy from ADC to Faust

// Arduino Setup
void setup(void) {
  // Open Serial
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);

  // Setup Faust volume control
  

  // Setup sine wave generator
  

  // Start I2S ADC (Input)
  auto cfg_adc = adc.defaultConfig(RX_MODE);
  cfg_adc.sample_rate = sample_rate ;
  cfg_adc.channels = 2; // Assuming 2-channel input from ADC
  cfg_adc.bits_per_sample = bits_per_sample;
  cfg_adc.port_no = 0;
  cfg_adc.pin_bck = 27;
  cfg_adc.pin_ws = 23;
  cfg_adc.pin_data = 18;
  config.pin_data_rx = 22;
  cfg_adc.pin_mck = 3;
  cfg_adc.is_master = true;
  cfg_adc.use_apll = true;
  adc.begin(cfg_adc);




  
   // Start I2S 2 (Output) ეს არის ფიქსირებული აუტი
  auto cfg_i2s2 = i2s_2.defaultConfig(TX_MODE);
  cfg_i2s2.sample_rate = cfg.sample_rate;
  cfg_i2s2.channels = 2; // Assuming 2-channel output
  cfg_i2s2.bits_per_sample = cfg.bits_per_sample;
  cfg_i2s2.port_no = 1;
  cfg_i2s2.pin_bck = 25;
  cfg_i2s2.pin_ws = 14;
  cfg_i2s2.pin_data = 15;
  cfg_i2s2.is_master = true;
  cfg_i2s2.use_apll = true;
  i2s_2.begin(cfg_i2s2);


  // Add I2S outputs to ChannelsSelectOutput
  out.addOutput(i2s_1, 0, 1);  // Output channels 0 and 1 to i2s_1
  out.addOutput(i2s_2, 2, 3);  // Output channels 2 and 3 to i2s_2
  out.begin(info4);  // Initialize the output with 4-channel info
}

// Arduino loop - copy sound to out
void loop() {
  copier.copy();
}

**Reply by pschatzmann** [reply]
Your sketch does not make any sense: you have only 2 I2S Ports available, so you can't use more then 2 I2SStream!

To have 1 I2S input and 2 I2S outputs you should use 1 in duplex mode (RXTX_MODE) and one for output only!

**Reply by daviti231** [reply]
Sorry, for auto ADC where I am using shared pins, default config I use (RXTX_MODE), just copied by mistake

cfg_adc = adc.defaultConfig(RXTX_MODE);
cfg_adc.sample_rate = sample_rate ;
cfg_adc.channels = 2; // Assuming 2-channel input from ADC
cfg_adc.bits_per_sample = bits_per_sample;
cfg_adc.port_no = 0;
cfg_adc.pin_bck = 27;
cfg_adc.pin_ws = 23;
cfg_adc.pin_data = 18;
config.pin_data_rx = 22;
cfg_adc.pin_mck = 3;
cfg_adc.is_master = true;
cfg_adc.use_apll = true;
adc.begin(cfg_adc);

**Reply by pschatzmann** [reply]
The rx pin does not point to the right structure! I also think out is wrong: it should be 2 channels where you output 0 and 1 to both

**Reply by daviti231** [reply]
I have tested on config.pin_data_rx = 19; as well but the same roblem is here

**Reply by pschatzmann** [reply]
It should be cfg_adc.cfg_adc !

**Reply by daviti231** [reply]
several day of working, I finalized my code but some issue still remains.
actualy my configuration is, to receive analo signal throgurh i2s-ADC and send it to two external HiLetgo PCM5102, thrgouh faust DSP, as As I am using faust as a filter for diferent purposes, in this particular case I wan to receive two high pass signal and two lowpass signal.
it seems it works when only activated out.begin(info); with two channel AudioInfo info(sample_rate, 2, bit_depth);  but when I am starting to activate out.begin(info4); with configuration :  AudioInfo info4(sample_rate, 4, bit_depth); it is giving noise sound on output, with two channel it works perfectly.  
please advice what I am doing wrong. I have tested this code with SineWaveGenerator and it work perfectly I can contrlo two channel separatelly and two channel separatelly, just in ADC is created problem.
 I also attach faust diagram 

#include "AudioTools.h"
#include "AudioLibs/AudioFaust.h"
#include "untitled11.h"

const uint16_t sample_rate = 22050;
const uint8_t bit_depth = 16;

AudioInfo info(sample_rate, 2, bit_depth);
AudioInfo info4(sample_rate, 4, bit_depth);

I2SStream i2s;
I2SStream out_2;
ChannelsSelectOutput out;
mydsp dsp;  // FAUST DSP object
FaustStream<mydsp> faust(out);  // FAUST stream using the output channel selector

StreamCopy copier(faust, i2s); //

void setup(void) {  
  // Open Serial
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning); 

  // Setup Faust volume control
  auto cfg = faust.defaultConfig();
  faust.begin(cfg);
  faust.setLabelValue("xma1", 0);  // volume for first DAC
  faust.setLabelValue("xma2", -10);  // volume for second dac
  faust.setLabelValue("SUBhighpass", 70);
  faust.setLabelValue("SUBlowpas", 2000);
  faust.setLabelValue("Tviteri_highpass", 2000);


  // Start I2S input
  Serial.println("Starting I2S...");

  auto config = i2s.defaultConfig(RXTX_MODE);
  config.copyFrom(info); 
  config.i2s_format = I2S_STD_FORMAT;
  config.sample_rate = sample_rate;
  config.bits_per_sample = bit_depth;  
  config.channels = 2;
  config.port_no = 0;
  config.pin_ws = 23;
  config.pin_bck = 27;
  config.pin_data = 18;
  config.pin_data_rx = 19;
  config.is_master = true;
  config.use_apll = true;  // Use APLL for better clock stability
  config.pin_mck = 3;
  //config.buffer_size = 2048; // Increased buffer size for stability
  i2s.begin(config);

  // Start I2S output for second DAC
  auto cfg_out = out_2.defaultConfig(TX_MODE);
  cfg_out.copyFrom(info);
  cfg_out.i2s_format = I2S_STD_FORMAT;
  cfg_out.bits_per_sample = bit_depth;
  cfg_out.sample_rate = sample_rate;
  cfg_out.channels = 2;
  cfg_out.is_master = true;
  cfg_out.port_no = 1;
  cfg_out.pin_bck = 14;
  cfg_out.pin_ws = 15;
  cfg_out.pin_data = 22;
  cfg_out.use_apll = true;
  cfg_out.pin_mck = -1;
  //cfg_out.buffer_size = 2048; // Ensure this is set correctly
  out_2.begin(cfg_out);

  // Configure output channels
  out.addOutput(i2s, 0, 1);
  out.addOutput(out_2, 2, 3);

  
  out.begin(info);
  //out.begin(info4);

  Serial.println("I2S started...");
}

void loop() {
  copier.copy();
}


![faust configuraion](https://github.com/pschatzmann/arduino-audio-tools/assets/137186205/9b975125-b16c-4fc0-8cb3-418b73f9b742)

**Reply by pschatzmann** [reply]
your diagram does not match the sketch and your sketch contains the following errors:
- your input to ChannelsSelectOutput is 2 channels and not 4 so using info4 is just wrong: just use info here as well
- the output definition is therefore wrong as well because you don't have 4 channels so you need to define the output of your 2 channels:

So  send the same output to both i2s channels
```
out.addOutput(i2s, 0, 1);
out.addOutput(out_2, 0, 1);
```

or split up the left to one i2s and right to the other i2s channel

```
out.addOutput(i2s, 0, 0);
out.addOutput(out_2, 1, 1);
```

**Reply by daviti231** [reply]
it works, just betveen channels is not good searation, is there any solution for  it?

**Reply by pschatzmann** [reply]
I don't understand: what do you mean with channel separation ?

**Reply by daviti231** [reply]
I mean, when sending stereo signal, left channnel signal is lisening on right channel signal and vise versa, but with very low level

**Reply by daviti231** [reply]
I want to ask, if there is posibility to Converting the Number of Input with ChannelFormatConverterStreamT so that, 2 input channel from ADC become 4 channel and than stream to 4 channel output

**Reply by pschatzmann** [reply]
> I mean, when sending stereo signal, left channnel signal is lisening on right channel signal and vise versa, but with very low level
That can't be: in the processing they are strictly seperated: I guess this must be a problem with your audio source

You can increase the channels: however this might not get you what you want. If I remember right the logic is that if you increase the channels more then available, the last channel is just copied to the additional ones. So you might need to do some additional processing there...
